erpnext/manufacturing/doctype/bom/bom.py

```diff
@@ -103,25 +104,27 @@ class BOM(WebsiteGenerator):
 	)
 
 	def autoname(self):
-		names = frappe.db.sql_list("""select name from `tabBOM` where item=%s""", self.item)
-
-		if names:
-			# name can be BOM/ITEM/001, BOM/ITEM/001-1, BOM-ITEM-001, BOM-ITEM-001-1
+		existing_boms = frappe.get_all("BOM", filters={"item": self.item}, pluck="name")
+		if existing_boms:
+			index = self.get_next_version_index(existing_boms)
+		else:
+			index = 1
 
-			# split by item
-			names = [name.split(self.item, 1) for name in names]
-			names = [d[-1][1:] for d in filter(lambda x: len(x) > 1 and x[-1], names)]
+		prefix = self.doctype
+		suffix = "%.3i" % index  # convert index to string (1 -> "001")
+		bom_name = f"{prefix}-{self.item}-{suffix}"
 
-			# split by (-) if cancelled
-			if names:
```

Comment:
@Alchez new logic for versioning doesn't consider cancelled items (check last test case I added which fails)